### PR TITLE
Introduce support for MongoDB service binding

### PR DESCRIPTION
**Pull request by sgahlot**
Initial codebase for adding the Service Binding for MongoDB.

### CODE REVIEW

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 1]
```diff
@@ -371,4 +368,20 @@ HealthBuildItem addHealthCheck(MongoClientBuildTimeConfig buildTimeConfig) {
...
+                            MongoServiceBindingConverter.class.getName()));
+        }
+    }
+
+    // TODO(SG): do we need this instead of the other registerServiceBinding?
+    @BuildStep
+    RunTimeConfigurationSourceValueBuildItem registerServiceBinding(KubernetesServiceBindingRecorder recorder,
+            KubernetesServiceBindingConfig config) {
+        return new RunTimeConfigurationSourceValueBuildItem(recorder.configSources(config));
+    }
```
Comment (geoand) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:386]: This should not be needed as the `quarkus-kubernetes-service-binding` creates them. From your POV, all that is needed is to register `io.quarkus.kubernetes.service.binding.runtime.ServiceBindingConverter`.
That said, I think you are missing a `META-INF/services/io.quarkus.kubernetes.service.binding.runtime.ServiceBindingConverter` like https://github.com/quarkusio/quarkus/blob/2.1.0.Final/extensions/kafka-client/runtime/src/main/resources/META-INF/services/io.quarkus.kubernetes.service.binding.runtime.ServiceBindingConverter

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingConverter.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    private static final String DB_PASSWORD = "password";
+    private static final String DB_HOST = "host";
+    private static final String DB_PORT = "port";
+    private static final String DB_DATABASE = "database";
+    private static final String DB_PREFIX = "mongodb+srv://";
+    private static final String DB_DEFAULT_OPTIONS = "?retryWrites=true&w=majority";
+
+    // Quarkus properties
+    private static final String QUARKUS_DS_USERNAME = "quarkus.datasource.username";
+    private static final String QUARKUS_DS_PASSWORD = "quarkus.datasource.password";
```
Comment (geoand) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingConverter.java:33]: These properties are specific to relational database handling and should not be set at all for Mongo. The Mongo integration adds these to the connection string, see https://quarkus.io/guides/all-config#quarkus-mongodb-client_quarkus.mongodb.connection-string

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingRecorder.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import org.eclipse.microprofile.config.spi.ConfigSourceProvider;
+
+// TODO(SG): do we need this class?
+@Recorder
+public class MongoServiceBindingRecorder extends KubernetesServiceBindingRecorder {
+    @Override
+    public RuntimeValue<ConfigSourceProvider> configSources(KubernetesServiceBindingConfig kubernetesServiceBindingConfig) {
+        return super.configSources(kubernetesServiceBindingConfig);
+    }
+}
```
Comment (geoand) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingRecorder.java:16]: This should not be needed either

**Review by geoand — COMMENTED** [review]
Nice work!

I've added some comments. Once they are addressed and an integration test is added, this should be good to go

File: extensions/mongodb-client/deployment/pom.xml [hunk 4]
```diff
@@ -34,6 +34,12 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-mongodb-client</artifactId>
         </dependency>
+        <!-- TODO(SG): We may not need this - depending on whether we need to use the kubernetes-service-binding code in Processor or not -->
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-kubernetes-service-binding</artifactId>
+            <optional>true</optional>
+        </dependency>
```
Comment (geoand) [inline extensions/mongodb-client/deployment/pom.xml:42]: Best leave it in to make sure that Maven builds the modules in the correct order

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java (hunk 1 above)
Comment (sgahlot) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:386]: I thought I added it in (at the very end of this page)

Comment (geoand) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:386]: Good point, it was too small in the UI so I missed it. In any case, the rest of the comment stands

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 5]
```diff
@@ -45,20 +48,12 @@
 import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.annotations.Weak;
-import io.quarkus.deployment.builditem.AdditionalIndexedClassesBuildItem;
-import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
-import io.quarkus.deployment.builditem.ExtensionSslNativeSupportBuildItem;
-import io.quarkus.deployment.builditem.FeatureBuildItem;
-import io.quarkus.deployment.builditem.SslNativeConfigBuildItem;
+import io.quarkus.deployment.builditem.*;
```
Comment (geoand) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:51]: Please don't use star imports

Comment (sgahlot) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:51]: I'll change my IDE settings.

**Comment by geoand** [comment]
Can you please squash the commits and rebase onto `main`?

Thanks

**Comment by geoand** [comment]
Seems like there are still multiple commits. It's would be great to have them squashed.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingConverter.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+ * <li>w=majority</li>
+ * </ul>
+ *
+ */
+public class MongoServiceBindingConverter implements ServiceBindingConverter {
+    private static final Logger LOGGER = Logger.getLogger(MongoServiceBindingConverter.class);
+
+    private static final String BINDING_TYPE = "mongodb";
+    public static final String BINDING_CONFIG_SOURCE_NAME = BINDING_TYPE + "-k8s-service-binding-source";
+    public static final String MONGO_DB_CONNECTION_STRING = "MONGODB_CONNECTION_STRING";
```
Comment (geoand) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingConverter.java:47]: This should be the Quarkus property, no?

**Comment by sgahlot** [comment]
@geoand I've squashed the commits for rebasing onto `main`, but not 100% sure if I did what you wanted. I've pushed that change to my branch. Here are the commands that I ran:
 `git rebase -i main` (from my branch)
 `git push --force origin <my_branch>`

I only see one commit when I run `git log`

**Comment by geoand** [comment]
GitHub is showing 7 commits, 3 of which are about this feature.
That's what I see in the UI at least

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingConverter.java (hunk 6 above)
Comment (sgahlot) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoServiceBindingConverter.java:47]: Yes, you're right. I took this from our hack code. The correct property should be `quarkus.mongodb.connection-string`. I'll fix this

**Comment by sgahlot** [comment]
George, do you know what command should I run to fix this? Earlier, we used to squash all the commits when merging the codebase but rebase also allowed me to squash my 2nd commit with the 1st one.
There must be something wrong from my end, I guess.

**Comment by geoand** [comment]
I usually just use IntelliJ's feature. See https://youtu.be/ah9ENkAYDrY for an example

**Comment by sgahlot** [comment]
I've now squashed the commits and also did rebase into `main`, but not too sure about that as I don't see any changes to be pushed for that op. Let me know if I need to do something regarding that (or not).

**Review by geoand — APPROVED** [review]
Thanks for this!

**Comment by geoand** [comment]
BTW, I pushed a commit into the PR that adds this module to one of the Mongo integration tests, thus verifying that everything is wired up correctly

**Comment by sgahlot** [comment]
Thanks for the Integration part. I had no clue about that.

**Pull request merged**